lib/src/nearest_point_on_line.dart

```diff
@@ -0,0 +1,215 @@
+        'location': super.location,
+      },
+    );
+  }
+}
+
+_Nearest _nearestPointOnLine(
+  LineString line,
+  Point point, [
+  Unit unit = Unit.kilometers,
+]) {
+  _Nearest? nearest;
+
+  num length = 0;
+
+  for (var i = 0; i < line.coordinates.length - 1; ++i) {
+    final startCoordinates = line.coordinates[i];
+    final stopCoordinates = line.coordinates[i + 1];
+
+    final startPoint = Point(coordinates: startCoordinates);
+    final stopPoint = Point(coordinates: stopCoordinates);
+
+    final sectionLength = distance(startPoint, stopPoint, unit);
+
+    final start = _Nearest(
+      point: startPoint,
+      distance: distance(point, startPoint, unit),
+      index: i,
+      location: length,
+    );
+
+    final stop = _Nearest(
+      point: stopPoint,
+      distance: distance(point, stopPoint, unit),
+      index: i + 1,
+      location: length + sectionLength,
+    );
+
+    final heightDistance = max(start.distance, stop.distance);
+    final direction = bearing(startPoint, stopPoint);
+
+    final perpendicular1 = destination(
+      point,
+      heightDistance,
+      direction + 90,
+      unit,
+    );
+
+    final perpendicular2 = destination(
+      point,
+      heightDistance,
+      direction - 90,
+      unit,
+    );
+
+    final intersectionPoint = intersects(
+      LineString.fromPoints(points: [perpendicular1, perpendicular2]),
+      LineString.fromPoints(points: [startPoint, stopPoint]),
+    );
+
+    _Nearest? intersection;
+
+    if (intersectionPoint != null) {
+      intersection = _Nearest(
+        point: intersectionPoint,
+        distance: distance(point, intersectionPoint, unit),
+        index: i,
+        location: length + distance(startPoint, intersectionPoint, unit),
+      );
+    }
+
+    if (nearest == null || start.distance < nearest.distance) {
+      nearest = start;
+    }
+
+    if (stop.distance < nearest.distance) {
+      nearest = stop;
+    }
+
+    if (intersection != null && intersection.distance < nearest.distance) {
+      nearest = intersection;
+    }
+
+    length += sectionLength;
+  }
+
+  /// A `LineString` is guaranteed to have at least two points and thus a
+  /// nearest point has to exist.
+
+  return nearest!;
+}
+
+_NearestMulti? _nearestPointOnMultiLine(
+  MultiLineString lines,
+  Point point, [
+  Unit unit = Unit.kilometers,
+]) {
+  _NearestMulti? nearest;
+
+  var globalIndex = 0;
+
+  for (var i = 0; i < lines.coordinates.length; ++i) {
+    final line = LineString(coordinates: lines.coordinates[i]);
+
+    final candidate = _nearestPointOnLine(line, point);
+    globalIndex += candidate.index;
+
+    if (nearest == null || candidate.distance < nearest.distance) {
+      nearest = _NearestMulti(
+        point: candidate.point,
+        distance: candidate.distance,
+        index: globalIndex,
+        localIndex: candidate.index,
+        location: candidate.location,
+        line: i,
+      );
+    }
+  }
+
+  return nearest;
+}
```

Comment:
Shouldn't `globalIndex` be a sum of the total line counts so far (instead of whatever `candidate.index` ends up being)?

```suggestion
_NearestMulti? _nearestPointOnMultiLine(
  MultiLineString lines,
  Point point, [
  Unit unit = Unit.kilometers,
]) {
  _NearestMulti? nearest;

  var globalIndex = 0;

  for (var i = 0; i < lines.coordinates.length; ++i) {
    final line = LineString(coordinates: lines.coordinates[i]);

    final candidate = _nearestPointOnLine(line, point);

    if (nearest == null || candidate.distance < nearest.distance) {
      nearest = _NearestMulti(
        point: candidate.point,
        distance: candidate.distance,
        index: globalIndex + candidate.index,
        localIndex: candidate.index,
        location: candidate.location,
        line: i,
      );
    }

    globalIndex += line.coordinates.length;
  }

  return nearest;
}
```